source/beamline/AMExclusiveStatesEnumeratedControl.cpp

```diff
@@ -0,0 +1,225 @@
+	// this control still reports as connected.
+
+	if (isConnected) {
+		QList<AMControl*> controls = indexControlMap_.values();
+
+		for (int i = 0, count = controls.count(); i < count && isConnected; i++) {
+			AMControl *control = controls.at(i);
+			isConnected = isConnected && control && control->isConnected();
+		}
+	}
+
+	setConnected(isConnected);
+}
+
+bool AMExclusiveStatesEnumeratedControl::addState(int index, const QString &stateName, double statusValue, AMControl *control, double controlTriggerValue)
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::addOption(index, stateName)) {
+		indexStatusMap_.insert(index, statusValue);
+		setControl(index, control);
+		indexTriggerMap_.insert(index, controlTriggerValue);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool AMExclusiveStatesEnumeratedControl::removeState(int index)
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::removeOption(index)) {
+		indexStatusMap_.remove(index);
+		removeControl(index);
+		indexTriggerMap_.remove(index);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool AMExclusiveStatesEnumeratedControl::clearStates()
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::clearOptions()) {
+		indexStatusMap_.clear();
+		clearControls();
+		indexTriggerMap_.clear();
+
+		result = true;
+	}
+
+	return result;
+}
+
+AMAction3* AMExclusiveStatesEnumeratedControl::createMoveAction(double optionIndex)
+{
+	AMAction3 *result = 0;
+
+	// Match the given index setpoint with the corresponding control setpoint.
+	// Create and return action that moves the control to the control setpoint.
+	// We can assume that the optionIndex given is valid, according to the
+	// validSetpoint() provided.
+
+	AMControl *control = indexControlMap_.value(optionIndex, 0);
+	double trigger = indexTriggerMap_.value(optionIndex, -1);
+
+	if (control)
+		result = AMActionSupport::buildControlMoveAction(control, trigger);
+
+	return result;
+}
+
+int AMExclusiveStatesEnumeratedControl::currentIndex() const
+{
+	// Initialize the new index to "Unknown".
+
+	int currentIndex = enumNames().indexOf("Unknown");
+
+	if (status_) {
+
+		// Identify the index corresponding to the control's current value.
+		// If there are multiple indices, pick the first. Ideally, this is
+		// an 'exclusive' states control--there shouldn't be multiple indices
+		// with the same state.
+
+		QList<int> indices = indicesWithStatus(status_->value());
+
+		if (!indices.isEmpty())
+			currentIndex = indices.first();
+	}
+
+	return currentIndex;
+}
+
+void AMExclusiveStatesEnumeratedControl::setControl(int index, AMControl *control)
+{
+	if (indexControlMap_.keys().contains(index) && indexControlMap_.value(index) != control)
+		removeControl(index);
+
+	indexControlMap_.insert(index, control);
+	addChildControl(control);
+}
+
+void AMExclusiveStatesEnumeratedControl::removeControl(int index)
+{
+	AMControl *toRemove = indexControlMap_.value(index, 0);
+
+	if (toRemove) {
+		removeChildControl(toRemove);
```

Comment:
There are no ownership issues here, right?  The `AMExclusiveStatesEnumeratedControl` is not responsible for the memory of the controls that it encapsulates?